cpp/src/spatial/hausdorff.cu

```diff
@@ -103,49 +135,92 @@ struct hausdorff_functor {
+
+    // ===== Make Space Size Iterator ==============================================================
+
+    auto d_space_offsets = cudf::column_device_view::create(space_offsets);
+
+    auto space_offset_iterator =
+      thrust::make_transform_iterator(thrust::make_counting_iterator<int32_t>(0),
+                                      size_from_offsets_functor{*d_space_offsets, xs.size()});
+
+    // ===== Make Hausdorff Accumulator ============================================================
 
-    // utilize one block per result (pair of spaces).
-    size_type num_blocks_x = min(result->size(), MAX_NUM_BLOCKS_X);
-    size_type num_blocks_y = ceil(result->size() / (float)MAX_NUM_BLOCKS_X);
+    auto num_distances = static_cast<int32_t>(num_points) * static_cast<int32_t>(num_points);
 
-    dim3 grid(num_blocks_x, num_blocks_y);
+    auto hausdorff_index_iter = thrust::make_transform_iterator(
+      thrust::make_counting_iterator<int32_t>(0),
+      haus_traversal_functor<decltype(space_offset_iterator)>{num_spaces,
+                                                              num_points,
+                                                              space_offsets.data<size_type>(),
+                                                              temp_space_lookup.data().get(),
+                                                              space_offset_iterator});
 
-    auto kernel = kernel_hausdorff<T>;
+    auto d_xs = cudf::column_device_view::create(xs);
+    auto d_ys = cudf::column_device_view::create(ys);
 
-    kernel<<<grid, MAX_NUM_THREADS_PER_BLOCK, 0, stream>>>(points_per_space.size(),
-                                                           xs.data<T>(),
-                                                           ys.data<T>(),
-                                                           space_offsets.data().get(),
-                                                           result->mutable_view().data<T>());
+    auto hausdorff_acc_iter = thrust::make_transform_iterator(
+      hausdorff_index_iter, hausdorff_index_to_acc_functor<T>{*d_xs, *d_ys});
 
-    CUDA_TRY(cudaGetLastError());
+    // ===== Materialize ===========================================================================
+
+    std::unique_ptr<cudf::column> result = make_column<T>(num_results, stream, mr);
+
+    auto result_temp      = rmm::device_buffer(sizeof(hausdorff_acc<T>) * num_results);
+    auto result_temp_iter = static_cast<hausdorff_acc<T>*>(result_temp.data());
+
+    auto scatter_map = thrust::make_transform_iterator(
+      hausdorff_index_iter, [] __device__(hausdorff_index acc) { return acc.result_idx; });
+
+    // the following output iterator and `inclusive_scan_by_key` could be replaced by a
+    // reduce_by_key, if it supported noncommutative operators.
+
+    auto scatter_out = make_scatter_output_iterator(result_temp_iter, scatter_map);
+
+    thrust::inclusive_scan_by_key(rmm::exec_policy(stream)->on(stream),
```

Comment:
Add a comment explaining what is happening.

Reply:
I removed this entirely in favor of an `upper_bound`, which is called once per point to determine which to which space that point belongs. The performance of that call scales with the number of spaces, not the number of points.

Reply:
The advantage of the `upper_bound` approach is that it does not require an extra `num_points * sizeof(int32_t)` bytes of memory. Since this O(n^2) approach's advantage is reduced memory usage, and not necessarily performance, I thought it appropriate to reduce the memory usage further. For 32768 points divided up in to 16384 spaces, the performance is comparable (presumably due to the overhead of the additional memory allocation and two kernel launches required by the previous implementation).

Reply:
Nice. I love upper_bound/lower_bound.